### ISSUE TITLE
storage: rebalancing - updates to positive constraints

### ISSUE THREAD

**Issue by BramGruneir** [cockroachdb/cockroach#14163]
There's some discussion to be had on exactly how we handle positive constraints and how we pick stores to add and remove replicas from will need to be updated with the way in which we handle them.

They were originally defined here:  https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/expressive_zone_config.md 
In the RFC, it specifies that we will always try to accept a positive constraint when possible, but there's no discussion on what the proper handling of multiple positive constraints are or how they should interact with required and prohibitive constraints.  

Some options for a single positive constraint:
- All - A single positive constraint will always move as many replicas onto available stores that have the desired attribute.  
- One - A single positive constraints will always more a single replica to a store with the desired attribute, all other replicas will remain on replicas without the constraint. 

Some options for two positive constraints:
- All Overlapping - if there are stores in which both constraints are met, move all replicas to those.  All other replicas should show a preference for stores with one of the two positive constraints. 
- All Non-overlapping - Freely choose between all nodes that contain at least one positive constraint. Note that this does not guarantee that all positive constraints are met.
- One Overlapping - A single replica will be present on the store that contains both, and this store will take precedence over any store that matches only a single positive constraint.
- One Non-overlapping - For each positive constraint, ensure at least one replica is located on each type of positive constraint.

Open questions: What should we do if there are more positive, non-overlapping constraints than replicas?

As for interaction with required/prohibitive constraints, I think those should always take precedence over positive constraints and we will not actively move replicas onto stores without those harder constraints being met.

**Comment by a-robinson** [comment]
I just want to +1 this issue for 1.2, since I do think that some middle ground between required and prohibited is going to be needed. The very old zone configs that let you manually specify a locality for each replica of a range were actually pretty useful, even if they had their issues.

**Comment by dianasaur323** [comment]
I am onboard with getting this issue in for 1.2. I view it as essential to our partitioning story, so we should address this. Are there any other concerns you have related to zone configs that you want to bubble up while we are at it?

**Comment by cuongdo** [comment]
@benesch since you're already working on partitioning, could you take a look for 1.2?

**Comment by a-robinson** [comment]
> Are there any other concerns you have related to zone configs that you want to bubble up while we are at it?

@dianasaur323 the other big one that comes to mind and that I don't know of an issue for is [moving some control of zone configs into SQL](https://cockroachlabs.slack.com/archives/G2N3Q10CA/p1505307467000446), but you're probably already aware of that.

**Comment by dianasaur323** [comment]
@a-robinson Yep, definitely aware of that at this point. My understanding was that we won't get to that in our first iteration of partitioning.

**Comment by danhhz** [comment]
We should consider adding "majority" to the "all" and "one" dimension discussed above.

That said, if this is getting scoped for 1.2 because of partitioning, it's in need of some additional customer validation. I'm not yet convinced the above description covers everything (one of the partitioning customers discusses wanting what amounts to replica level required constraints in https://github.com/cockroachdb/cockroach/issues/16978) and I'd hate to do a bunch of work on positive constraints if what we really need is to reintroduce per-replica required constraints. Reassigning to Andy, who has been handling talking to customers about their partitioning needs

**Comment by cuongdo** [comment]
cc @nvanbenschoten

**Comment by petermattis** [comment]
@a-robinson Given the recent updates to zone config expressiveness, should we consider this done?
